lantiq: xrx200: enable use of baby jumbo frames

tomjnixon opened the pull request:
xrx200 MTU is reduced so that the maximum mtu setting works correctly.

The maximum packet size accepted by the switch is increased to match the maximum mtu.

The motivation is to allow an ethernet MTU of 1508, which can carry a PPPoE session with an MTU of 1500 as per rfc4638. Currently if you use a lantiq device as an ADSL/VDSL modem only, you have to live with a MTU of 1492, which is less than ideal.

I'd like to send this upstream; haven't done that yet though.

EDIT: the following discussion is now out of date, see [this thread](https://github.com/openwrt/openwrt/pull/4353#discussion_r668331201).

This slightly changes the switch configuration, so could affect performance -- I'll run some tests on this in the next few days. So far I've tested both patches on a netgear DM200 and a homehub v5a with maximum-MTU pings (`ping -s 1480`) and regular network traffic.

The thing I'm not 100% sure about with this is that the switch ports are all set to work with a 1508-byte mtu, because (AFAIK) there's only one setting for this, and a per-port enable. The DSA slave device doesn't reject packets over the configured MTU limit, so you can have an interface with a configured MTU of 1500 which passes 1508-byte packets fine.

I think this should be ok -- in other cases this seems unavoidable given how VLAN tags are handled -- but it's not totally clear to me.

tomjnixon committed:
lantiq: xrx200: enable use of baby jumbo frames

xrx200 max MTU is reduced so that it works correctly when set to the
max, and the max MTU of the switch is increased to match.

In 5.10, the switch driver now enables non-standard MTUs on a per-port
basis, with the overall frame size set based on the cpu port. When the
MTU is not used, this should have no effect. The maximum packet size is
limited as large packets cause the switch to lock up.

0702-net-lantiq-add-support-for-jumbo-frames.patch comes from net-next
commit 998ac358019e491217e752bc6dcbb3afb2a6fa3e.

In 5.4, all switch ports are configured to accept the max MTU, as 5.4
does not have port_max_mtu/port_change_mtu callbacks.

Signed-off-by: Thomas Nixon <tom@tomn.co.uk>

abajk commented on target/linux/lantiq/patches-5.4/0702-net-lantiq-enable-baby-jumbo-frames-on-xrx200.patch:
```diff
@@ -0,0 +1,61 @@
+From 9fd8a972e4b9691797e62ce27687ab9f94881b2f Mon Sep 17 00:00:00 2001
```
Patch for 5.4 differs from that for kernel 5.10. It's a mistake?

tomjnixon replied:
Yes, the way MTUs are handled for DSA slave interfaces is different in these two versions.

In 5.4 max_mtu is hard-coded to ETH_MAX_MTU, and it's up to the user to set the mtu correctly.

5.10 has port_change_mtu and port_max_mtu callbacks and makes sure that the master interface has a big enough mtu for all the slave interfaces when the slave interface MTU is changed.

abajk commented:
Overall it looks good to me.

tomjnixon commented:
> This switch supports Jumbo frame up to 9k[1].

Yeah. My assumption was that nobody would want to use jumbo frames on this hardware, except for rfc4638. Supporting 9k frames in the ethernet driver would end up being a bit more complex, because we would probably want to dynamically reallocate the receive buffers when the MTU is changed, and it didn't seem worth the effort.

Thanks for having a look.

tomjnixon commented:
Did some performance tests on a homehub v5a, seems fine.

I saw a few times when traffic which should have stayed in the switch was going through the CPU port; not sure exactly how to debug that though, and it doesn't seem related to this patch.

<details>
  <summary>full test results</summary>

## with mtu patches

### to CPU port with no bridge
    iperf3 -c 192.168.1.1 -u -b 0
    [  5]   0.00-10.00  sec   131 MBytes   110 Mbits/sec  0.147 ms  723456/818381 (88%)  receiver
    
    iperf3 -c 192.168.1.1 -u -b 0 -R
    [  5]   0.00-10.00  sec   106 MBytes  89.1 Mbits/sec  0.009 ms  0/76957 (0%)  receiver
    
    iperf3 -c 192.168.1.1 -R -P2
    [SUM]   0.00-10.00  sec   315 MBytes   264 Mbits/sec                  receiver
    
    iperf3 -c 192.168.1.1 -P2
    [SUM]   0.00-10.01  sec   369 MBytes   309 Mbits/sec                  receiver

### to CPU port with bridge

    iperf3 -c 192.168.1.1 -u -b 0
    [  5]   0.00-10.00  sec   128 MBytes   108 Mbits/sec  0.231 ms  725229/818211 (89%)  receiver
    
    iperf3 -c 192.168.1.1 -u -b 0 -R
    [  5]   0.00-10.00  sec  90.2 MBytes  75.7 Mbits/sec  0.012 ms  0/65345 (0%)  receiver
    
    iperf3 -c 192.168.1.1 -R -P2
    [SUM]   0.00-10.00  sec   257 MBytes   215 Mbits/sec                  receiver
    
    iperf3 -c 192.168.1.1 -P2
    [SUM]   0.00-10.00  sec   236 MBytes   198 Mbits/sec                  receiver

### through switch

    iperf3 -c 192.168.1.3 -u -b 0 -P4
    [SUM]   0.00-10.00  sec  1.01 GBytes   870 Mbits/sec  0.060 ms  0/751272 (0%)  receiver
    [SUM]   0.00-10.00  sec  1.05 GBytes   899 Mbits/sec  0.118 ms  42519/819004 (5.2%)  receiver
    
    iperf3 -c 192.168.1.3 -u -b 0 -R -P4
    [SUM]   0.00-10.00  sec  1.01 GBytes   868 Mbits/sec  0.064 ms  0/749420 (0%)  receiver
    [SUM]   0.00-10.00  sec  1.01 GBytes   871 Mbits/sec  0.029 ms  51/751557 (0.0068%)  receiver
    
    iperf3 -c 192.168.1.3 -R -P4
    [SUM]   0.00-10.00  sec  1.10 GBytes   941 Mbits/sec                  receiver
    
    iperf3 -c 192.168.1.3 -P4
    [SUM]   0.00-10.00  sec  1.09 GBytes   934 Mbits/sec                  receiver

## without mtu patches

### to CPU port with no bridge

    iperf3 -c 192.168.1.1 -u -b 0
    [  5]   0.00-10.20  sec   127 MBytes   105 Mbits/sec  0.104 ms  726818/818980 (89%)  receiver
    
    iperf3 -c 192.168.1.1 -u -b 0 -R
    [  5]   0.00-10.00  sec   121 MBytes   102 Mbits/sec  0.018 ms  0/87701 (0%)  receiver
    
    iperf3 -c 192.168.1.1 -R -P2
    [SUM]   0.00-10.00  sec   325 MBytes   272 Mbits/sec                  receiver
    
    iperf3 -c 192.168.1.1 -P2
    [SUM]   0.00-10.00  sec   361 MBytes   303 Mbits/sec                  receiver

### to CPU port with bridge

    iperf3 -c 192.168.1.1 -u -b 0
    [  5]   0.00-10.21  sec   123 MBytes   101 Mbits/sec  0.099 ms  730041/818901 (89%)  receiver
    
    iperf3 -c 192.168.1.1 -u -b 0 -R
    [  5]   0.00-10.00  sec  89.9 MBytes  75.4 Mbits/sec  0.007 ms  0/65120 (0%)  receiver
    
    iperf3 -c 192.168.1.1 -R -P2
    [SUM]   0.00-10.00  sec   252 MBytes   211 Mbits/sec                  receiver
    
    iperf3 -c 192.168.1.1 -P2
    [SUM]   0.00-10.00  sec   225 MBytes   189 Mbits/sec                  receiver

### through switch

    iperf3 -c 192.168.1.3 -u -b 0 -P4
    [SUM]   0.00-10.00  sec   716 MBytes   601 Mbits/sec  0.055 ms  298038/816792 (36%)  receiver
    [SUM]   0.00-10.00  sec   970 MBytes   814 Mbits/sec  0.106 ms  116424/819028 (14%)  receiver
    [SUM]   0.00-10.00  sec   878 MBytes   736 Mbits/sec  0.090 ms  180180/815723 (22%)  receiver
    [SUM]   0.00-10.00  sec   878 MBytes   737 Mbits/sec  0.130 ms  182960/818925 (22%)  receiver
    [SUM]   0.00-10.00  sec  1.10 GBytes   948 Mbits/sec  0.054 ms  0/818396 (0%)  receiver
    
    iperf3 -c 192.168.1.3 -u -b 0 -R -P4
    [SUM]   0.00-10.00  sec  1.02 GBytes   874 Mbits/sec  0.060 ms  0/754393 (0%)  receiver
    
    iperf3 -c 192.168.1.3 -R -P4
    [SUM]   0.00-10.00  sec  1.10 GBytes   941 Mbits/sec                  receiver
    
    iperf3 -c 192.168.1.3 -P4
    [SUM]   0.00-10.00  sec  1.09 GBytes   934 Mbits/sec                  receiver

</details>

LGA1150 commented on target/linux/lantiq/patches-5.10/0702-net-lantiq-enable-baby-jumbo-frames-on-xrx200.patch:
```diff
@@ -0,0 +1,99 @@
++
++static int gswip_port_change_mtu(struct dsa_switch *ds, int port, int new_mtu)
++{
++	return 0;
```
You should only write to `GSWIP_MAC_FLEN` in `.port_change_mtu`, instead of hardcoding it elsewhere.

Please consider upstreaming this too.

tomjnixon replied:
Thanks for taking a look. My reason for not doing this is that there is only one FLEN register shared between all switch ports, which also needs to account for the DSA tag on the CPU port. Therefore when the MTU of a port is set to anything other than 1500, it is always going to be able to accept larger packets than intended.

Since this doesn't seem to have any negative effects, we might as well just set all ports to support the maximum size we want, and save some code in the process. Does this make sense, or am i being lazy?

I'll have a go and see how it looks anyway.

LGA1150 replied:
You can take a look at how I implement this for MT7530 switch: torvalds/linux@9470174e7581e75a8ebd78964997314dfc2e706c, remember to test the edge cases!
And if the global `GSWIP_MAC_FLEN` register actually configures max RX length (MRU), you should enable `mtu_enforcement_ingress` in setup function: torvalds/linux@771c8901568dd8776a260aa93db41be88a60389e

tomjnixon replied:
Thanks, that looks a bit cleaner than what i was thinking about -- I'll give it a go later.

tomjnixon replied:
Latest push should be as per your suggestion, thanks for that.

The 5.4 and 5.10 patches now operate quite differently; I'm not exactly sure how is best to handle that. On the other hand, the 5.10 patch now doesn't really change anything if the user doesn't change the MTU, so it's a bit less scary.

LGA1150 replied:
If the switch supports 9K jumbo frames, you should lift the limit.

tomjnixon replied:
Sorry i meant to mention that; that was the highest i was able to test so far. I'll have another go later; it's a bit tricky, because the ethernet driver doesn't seem to work right with larger frames.

tomjnixon replied:
Updated to use the max packet length from the DM200 GPL dump. I can't test above 9K, but the switch still passes 9K packets when set to the maximum (with the ethernet driver modified to allow higher MTUs), so it should be fine.

In practice it's always limited by the max MTU of the ethernet driver, so this is only really relevant for someone working on modifying the ethernet driver to support larger packets.

LGA1150 replied:
Does the CPU port always need `GSWIP_MAC_CTRL_2_MLEN`, even if all user ports MTU <= 1492?
If it does not, you can simplify the function
```c
if (port == cpu_port) {
	new_mtu += 8;
	gswip_switch_w(priv, VLAN_ETH_HLEN + new_mtu + ETH_FCS_LEN,
		       GSWIP_MAC_FLEN);
}

if (new_mtu > ETH_DATA_LEN)
	gswip_switch_mask(priv, 0, GSWIP_MAC_CTRL_2_MLEN,
			  GSWIP_MAC_CTRL_2p(port));
else
	gswip_switch_mask(priv, GSWIP_MAC_CTRL_2_MLEN, 0,
			  GSWIP_MAC_CTRL_2p(port));
```

tomjnixon replied:
Good point, I've made that change.

Also changed `>` to `!=`, because MLEN/FLEN also work for reducing the maximum accepted  packet size.

jospezial commented:
Is this ready to merge?
Had no problems with this PR on XRX200 VGV7510KW22 (o2 Box 6431) for some weeks.

tomjnixon commented:
> Is this ready to merge?
> Had no problems with this PR on XRX200 VGV7510KW22 (o2 Box 6431) for some weeks.

From my point of view, yes -- it should just need attention from a maintainer.

Thanks for testing!

jospezial commented:
> > Is this ready to merge?
> > Had no problems with this PR on XRX200 VGV7510KW22 (o2 Box 6431) for some weeks.
> 
> From my point of view, yes -- it should just need attention from a maintainer.
> 
> Thanks for testing!
@hauke  @ynezz @mkresin please merge.

abajk commented:
@tomjnixon You should discuss your patch with the linux net maintainers.

Ref: https://lists.openwrt.org/pipermail/openwrt-devel/2021-April/034639.html

hauke commented on target/linux/lantiq/patches-5.10/0702-net-lantiq-enable-baby-jumbo-frames-on-xrx200.patch:
```diff
@@ -0,0 +1,125 @@
+ 
+ #define XRX200_GPHY_FW_ALIGN	(16 * 1024)
+ 
++#define GSWIP_MAX_PACKET_LENGTH	9600
```
The PHYs can handle up to 9990 bytes and the switch buffer manager up to 10240 bytes.
We should make sure to not send packets bigger than 9990 bytes.

tomjnixon replied:
Sorry for the delay. The 9600 value came from the DM200 GPL dump; I could increase this, but as written above I  can't test these larger sizes.

hauke commented on target/linux/lantiq/patches-5.10/0702-net-lantiq-enable-baby-jumbo-frames-on-xrx200.patch:
```diff
@@ -0,0 +1,125 @@
++	if (port == cpu_port) {
++		new_mtu += 8;
++		gswip_switch_w(priv, VLAN_ETH_HLEN + new_mtu + ETH_FCS_LEN,
++			       GSWIP_MAC_FLEN);
```
This is not configuring the MTU for the CPU port, but what the switch will handle internally.

tomjnixon replied:
The comment above aims to address this:

> cpu port always has maximum mtu of user ports, so use it to set switch frame size, including 8 byte special header

Does this make sense? Doing it this way, rather than finding the maximum port MTU was suggested by  LGA1150 above.

xdarklight replied:
I think GSWIP is slightly different than MT7530 (which is what @LGA1150 worked on).
My understanding is that:
- both have a global register for the maximum MTU which is then valid across all ports. on GSWIP the register is GSWIP_MAC_FLEN
- in addition to that GSWIP can limit the frame length per port to 1518 + VLAN tag length (which is not possible on MT7530 from what I understand). on GSWIP the register is GSWIP_MAC_CTRL_2p(port)

Unfortunately I am not an expert on this topic. when in doubt: the netdev maintainers are very helpful and respond fast (I always get a reply within 24 hours)

hauke commented on target/linux/lantiq/patches-5.10/0702-net-lantiq-enable-baby-jumbo-frames-on-xrx200.patch:
```diff
@@ -0,0 +1,125 @@
++				  GSWIP_MAC_CTRL_2p(port));
++	else
++		gswip_switch_mask(priv, GSWIP_MAC_CTRL_2_MLEN, 0,
++				  GSWIP_MAC_CTRL_2p(port));
```
`GSWIP_MAC_CTRL_2_MLEN` has these options:
0: Normal (MTU of 1518 bytes ) VLAN tags are not included and can be added in addition.
1: Jumbo: the value from `GSWIP_MAC_FLEN` is used and it is with all VLAN tags. 

With `GSWIP_MAC_CTRL_2_LCHKL` you can also deactivate the length check. bit (2)

tomjnixon replied:
Thanks, I hadn't seen that. Do you think it would make more sense to disable the length check on the CPU port so that the special header does not affect the maximum packet size on user ports with non-standard MTUs, as it does currently?

xdarklight replied:
from what I understand DSA includes the tag header size ("tag protocol overhead") in the MTU calculation for the CPU port:
https://elixir.bootlin.com/linux/v5.15-rc2/source/net/dsa/slave.c#L1640

hauke replied:
I would prefer if we still check the MTU on the CPU port we should at least make sure the packets are fitting into the DMA descriptors, I do not know what the hardware will do if the DMA descriptors are too small.

tomjnixon replied:
@xdarklight The tag overhead is not included in the CPU port MTU, but it is included in the master port MTU, so i think this logic is correct.

Adding a printk to gswip_port_change_mtu:

```c
printk("gswip_port_change_mtu %d %d %d\n", port, cpu_port, new_mtu);
```

results in this on boot up:
```
[   10.260315] gswip_port_change_mtu 6 6 1500
[   10.262955] gswip_port_change_mtu 0 6 1500
[   10.283266] gswip_port_change_mtu 1 6 1500
[   10.303300] gswip_port_change_mtu 2 6 1500
[   10.322546] gswip_port_change_mtu 4 6 1500
[   10.342211] gswip_port_change_mtu 5 6 1500
```

And this when running `ip link set dev lan1 mtu 1508`:
```
[  144.866115] gswip_port_change_mtu 6 6 1508
[  144.868820] gswip_port_change_mtu 4 6 1508
[  144.873023] gswip_port_change_mtu 2 6 1508
[  144.876964] gswip_port_change_mtu 1 6 1508
[  144.881200] gswip_port_change_mtu 0 6 1508
```

tomjnixon replied:
@hauke

> I would prefer if we still check the MTU on the CPU port we should at least make sure the packets are fitting into the DMA descriptors, I do not know what the hardware will do if the DMA descriptors are too small.

The DSA infrastructure doesn't allow the slave device mtu to be set higher than the maximum mtu of the master device (minus the overhead), so with the current patch I don't think this is a problem.

Unfortunately it is possible to reduce the master MTU below that required by the slaves, e.g.:

```bash
ip link set dev lan1 mtu 1508 # master: 1516, slaves: 1508
ip link set dev eth0 mtu 1500 # master: 1500, slaves: 1508
```

In theory this might cause an issue, but the [xrx200 jumbo frame patch](https://git.kernel.org/pub/scm/linux/kernel/git/netdev/net-next.git/commit/?id=998ac358019e491217e752bc6dcbb3afb2a6fa3e) from @abajk never reduces the buffer sizes, so it should be OK. I've tried this and couldn't see any issues, but maybe there's a better way that I'm missing?

abajk replied:
> In theory this might cause an issue, but the [xrx200 jumbo frame patch](https://git.kernel.org/pub/scm/linux/kernel/git/netdev/net-next.git/commit/?id=998ac358019e491217e752bc6dcbb3afb2a6fa3e) from @abajk never reduces the buffer sizes, so it should be OK. I've tried this and couldn't see any issues, but maybe there's a better way that I'm missing?

My patch does not reallocate buffers if the new MTU is less than the old one. Having a larger buffer, the driver can receive a shorter packet. If the driver receives a packet, the next buffer buffer will be smaller. After 64 packets are received, all buffers will be smaller.

abajk replied:
> OK. I've tried this and couldn't see any issues, but maybe there's a better way that I'm missing?

It means that large packets on your device do not fhangs the switch?

tomjnixon replied:
> > In theory this might cause an issue, but the [xrx200 jumbo frame patch](https://git.kernel.org/pub/scm/linux/kernel/git/netdev/net-next.git/commit/?id=998ac358019e491217e752bc6dcbb3afb2a6fa3e) from @abajk never reduces the buffer sizes, so it should be OK. I've tried this and couldn't see any issues, but maybe there's a better way that I'm missing?
> 
> My patch does not reallocate buffers if the new MTU is less than the old one. Having a larger buffer, the driver can receive a shorter packet. If the driver receives a packet, the next buffer buffer will be smaller. After 64 packets are received, all buffers will be smaller.

Ah right, thanks for clarifying that.

> > OK. I've tried this and couldn't see any issues, but maybe there's a better way that I'm missing?
> 
> It means that large packets on your device do not fhangs the switch?

Sorry, i meant issues with reducing the master MTU below the mtu required by the slaves.

I was not testing with enough packets, but trying it again it just seems to drop packets, which should be fine.

I see the same issue as you with large packets hanging the switch -- assuming it is a problem with the switch, GSWIP_MAX_PACKET_LENGTH should probably be reduced.

abajk commented:
Hi Tomas,
> 
> 
> > This switch supports Jumbo frame up to 9k[1].
> 
> Yeah. My assumption was that nobody would want to use jumbo frames on this hardware, except for rfc4638. Supporting 9k frames in the ethernet driver would end up being a bit more complex, because we would probably want to dynamically reallocate the receive buffers when the MTU is changed, and it didn't seem worth the effort.

I wrote a patch that does this job[1]. Can you include my patch on this PR? 

1. https://github.com/abajk/linux/commit/868f8053a27349dd778693744138cdf4d00be28a

abajk commented:
I don't know why, but on BT Home Hub 5A it doesn't work for MTUs larger than 2510B. A ping greater than 2536B hangs the switch. I used my patch for lantiq_xrx200.c and yours for lantiq_gswip.c

xdarklight commented:
> I don't know why, but on BT Home Hub 5A it doesn't work for MTUs larger than 2510B. A ping greater than 2536B hangs the switch. I used my patch for lantiq_xrx200.c and yours for lantiq_gswip.c

I haven't tried this myself yet. How are you testing: from CPU port to any LAN/WAN port (or vice versa) or from any LAN/WAN port to another LAN/WAN port?
The most interesting registers are probably `GSWIP_MAC_FLEN` and `GSWIP_MAC_CTRL_2p(port)` (the latter for both affected ports)

abajk commented:
> > I don't know why, but on BT Home Hub 5A it doesn't work for MTUs larger than 2510B. A ping greater than 2536B hangs the switch. I used my patch for lantiq_xrx200.c and yours for lantiq_gswip.c
> 
> I haven't tried this myself yet. How are you testing: from CPU port to any LAN/WAN port (or vice versa) or from any LAN/WAN port to another LAN/WAN port?
> The most interesting registers are probably `GSWIP_MAC_FLEN` and `GSWIP_MAC_CTRL_2p(port)` (the latter for both affected ports)

The router hangs after sending a large packet from the LAN to the CPU and from the LAN port to another LAN port.

hauke commented:
It could be that the PPA engine which is on the RX part of the device and runs a firmware has some problems with bigger packets, I do not know much of its inner workings.

tomjnixon commented:
Some updates:

- Rebased onto the latest kernel versions. Would it make sense to forget about the 5.4 version of this, as the switch to 5.10 seems imminent, and it's not really possible to do it properly?

- Added the upstream mtu patch from @abajk; it seems to work well for me. My only complaint is that the 4 bytes of overhead do not come from the FCS (ETH_FCS_LEN), that is stripped out before the PMAC. Those bytes are required because the hardware can't always write to the last few bytes of the buffer, depending on how it's aligned. I checked, and packets with a bad FCS are dropped by the hardware in the current configuration.

- I found that the minimum MTU setting that would prevent crashes is 2526, so set `GSWIP_MAX_PACKET_LENGTH` to `2526 + VLAN_ETH_HLEN + ETH_FCS_LEN + 8` to enforce this. I'm not sure why I got a different number to you @abajk -- how are you testing it? I'm just setting the MTU and pinging with increasingly larger data sizes until it breaks. Interestingly, when this happens the device starts sending ethernet pause frames (i think after 16 further packets) -- maybe that hints at what's going wrong.

- @hauke wanted to make sure that frames from the switch aren't too big for the DMA descriptors. I found out what happens in this case: the packet gets put in two DMA descriptors, one with the SOP bit set, and another with the EOP bit set, so a patch that drops packets without SOP and EOP set is probably in order.

- There's a very slight issue with using the CPU port mtu to set FLEN: when setting a slave MTU, if the master device already had a large enough MTU then port_change_mtu is not called on the CPU port. This feels like it might not be intended; I'll ask on netdev.

abajk commented:
>     * I found that the minimum MTU setting that would prevent crashes is 2526, so set `GSWIP_MAX_PACKET_LENGTH` to `2526 + VLAN_ETH_HLEN + ETH_FCS_LEN + 8` to enforce this. I'm not sure why I got a different number to you @abajk -- how are you testing it? I'm just setting the MTU and pinging with increasingly larger data sizes until it breaks. Interestingly, when this happens the device starts sending ethernet pause frames (i think after 16 further packets) -- maybe that hints at what's going wrong.

With the current maximum value of MTU (2526), the switch will not hang. However, large packages are truncated. 

PC:
```
PING 192.168.1.1 (192.168.1.1) 2510(2538) bytes of data.
^C
--- 192.168.1.1 ping statistics ---
4 packets transmitted, 0 received, 100% packet loss, time 3060ms
```
Router:
```
root@OpenWrt:/# tcpdump -ni lan1
tcpdump: verbose output suppressed, use -v or -vv for full protocol decode
listening on lan1, link-type EN10MB (Ethernet), capture size 262144 bytes
08:55:39.538475 ARP, Request who-has 192.168.1.1 tell 192.168.1.10, length 46
08:55:39.538598 ARP, Reply 192.168.1.1 is-at 5c:b1:3e:45:b6:c2, length 28
08:55:59.997814 IP truncated-ip - 2 bytes missing! 192.168.1.10 > 192.168.1.1: ICMP echo request, id 3691, seq 1, length 2518
08:56:01.010122 IP truncated-ip - 2 bytes missing! 192.168.1.10 > 192.168.1.1: ICMP echo request, id 3691, seq 2, length 2518
08:56:02.034126 IP truncated-ip - 2 bytes missing! 192.168.1.10 > 192.168.1.1: ICMP echo request, id 3691, seq 3, length 2518
08:56:03.058103 IP truncated-ip - 2 bytes missing! 192.168.1.10 > 192.168.1.1: ICMP echo request, id 3691, seq 4, length 2518
```

abajk commented:
> It could be that the PPA engine which is on the RX part of the device and runs a firmware has some problems with bigger packets, I do not know much of its inner workings. If TX with bigger MTU works and only RX is broken this would point to the PPA engine.

@hauke You're right, only RX hangs with large packets. TX works up to 10kB.

nicefile reviewed (approved):
baby jumbo frames run on 7412 with rfc4638 setup  from July 21th . Tested VLAN bridge and non-VLAN bridge .

abajk commented:
@tomjnixon

>     * Rebased onto the latest kernel versions. Would it make sense to forget about the 5.4 version of this, as the switch to 5.10 seems imminent, and it's not really possible to do it properly?

Lantiq now uses kernel 5.10. You can drop the part for 5.4. 

>     * Added the upstream mtu patch from @abajk; it seems to work well for me. My only complaint is that the 4 bytes of overhead do not come from the FCS (ETH_FCS_LEN), that is stripped out before the PMAC. Those bytes are required because the hardware can't always write to the last few bytes of the buffer, depending on how it's aligned. 

This [Patch](https://git.kernel.org/pub/scm/linux/kernel/git/netdev/net.git/commit/?id=1488fc204568f707fe2a42a913788c00a95af30e) fixes this bug.


FYI. Some time ago I wrote a [patch](https://github.com/abajk/linux/commit/a33cd995ca6b5e139171ba624af61a459ebade58) to add support for SG DMA. It needs clean up. I will probably send it upstream early next week.

jospezial commented:
> This [Patch](https://git.kernel.org/pub/scm/linux/kernel/git/netdev/net.git/commit/?id=1488fc204568f707fe2a42a913788c00a95af30e) fixes this bug.

@tomjnixon Would you please cherry pick that?
https://github.com/torvalds/linux/commit/1488fc204568f707fe2a42a913788c00a95af30e
Will be part of 5.16-rc7.
@abajk When will we see your latest commits in stable (LTS) kernel?

jospezial commented:
> @tomjnixon Would you please cherry pick that?
> [torvalds/linux@1488fc2](https://github.com/torvalds/linux/commit/1488fc204568f707fe2a42a913788c00a95af30e)
@abajk
Does your patch need something else, we don't have in 5.10 yet?
```
  ccache_cc -Wp,-MMD,drivers/net/ethernet/.lantiq_xrx200.o.d -nostdinc -isystem /home/joerg/Downloads/openwrt6431/staging_dir/toolchain-mips_24kc_gcc-11.2.0_musl/lib/gcc/mips-openwrt-linux-musl/11.2.0/include -I./arch/mips/include -I./arch/mips/include/generated  -I./include -I./arch/mips/include/uapi -I./arch/mips/include/generated/uapi -I./include/uapi -I./include/generated/uapi -include ./include/linux/kconfig.h -include ./include/linux/compiler_types.h -D__KERNEL__ -DVMLINUX_LOAD_ADDRESS=0xffffffff80002000 -DLINKER_LOAD_ADDRESS=0x80002000 -DDATAOFFSET=0 -fmacro-prefix-map=./= -Wall -Wundef -Werror=strict-prototypes -Wno-trigraphs -fno-strict-aliasing -fno-common -fshort-wchar -fno-PIE -Werror=implicit-function-declaration -Werror=implicit-int -Werror=return-type -Wno-format-security -std=gnu89 -mno-check-zero-division -mabi=32 -G 0 -mno-abicalls -fno-pic -pipe -mno-branch-likely -msoft-float -DGAS_HAS_SET_HARDFLOAT -Wa,-msoft-float -ffreestanding -EB -fno-stack-check -Wa,-mno-fix-loongson3-llsc -march=mips32r2 -mtune=34kc -Wa,--trap -DTOOLCHAIN_SUPPORTS_VIRT -DTOOLCHAIN_SUPPORTS_XPA -DTOOLCHAIN_SUPPORTS_CRC -DTOOLCHAIN_SUPPORTS_DSP -DTOOLCHAIN_SUPPORTS_GINV -I./arch/mips/include/asm/mach-lantiq -I./arch/mips/include/asm/mach-lantiq/xway -I./arch/mips/include/asm/mach-generic -fno-asynchronous-unwind-tables -fno-delete-null-pointer-checks -Wno-frame-address -Wno-format-truncation -Wno-format-overflow -Wno-address-of-packed-member -O2 -fno-caller-saves -fno-allow-store-data-races -Wframe-larger-than=1024 -fstack-protector -Wimplicit-fallthrough -Wno-unused-but-set-variable -Wno-unused-const-variable -fomit-frame-pointer -g -femit-struct-debug-baseonly -fno-var-tracking -Wdeclaration-after-statement -Wvla -Wno-pointer-sign -Wno-stringop-truncation -Wno-zero-length-bounds -Wno-array-bounds -Wno-stringop-overflow -Wno-restrict -Wno-maybe-uninitialized -fno-strict-overflow -fno-stack-check -fconserve-stack -Werror=date-time -Werror=incompatible-pointer-types -Werror=designated-init -Wno-packed-not-aligned -fmacro-prefix-map=/home/joerg/Downloads/openwrt6431/build_dir/target-mips_24kc_musl=target-mips_24kc_musl  -ffunction-sections -fdata-sections  -DKBUILD_MODFILE='"drivers/net/ethernet/lantiq_xrx200"' -DKBUILD_BASENAME='"lantiq_xrx200"' -DKBUILD_MODNAME='"lantiq_xrx200"' -c -o drivers/net/ethernet/lantiq_xrx200.o drivers/net/ethernet/lantiq_xrx200.c
In file included from ./include/linux/skbuff.h:13,
                 from ./include/linux/if_ether.h:19,
                 from ./include/linux/etherdevice.h:20,
                 from drivers/net/ethernet/lantiq_xrx200.c:10:
drivers/net/ethernet/lantiq_xrx200.c: In function 'xrx200_buffer_size':
drivers/net/ethernet/lantiq_xrx200.c:111:56: error: 'XRX200_DMA_BURST_LEN' undeclared (first use in this function); did you mean 'XRX200_DMA_DATA_LEN'?
  111 |         return round_up(xrx200_max_frame_len(mtu), 4 * XRX200_DMA_BURST_LEN);
      |                                                        ^~~~~~~~~~~~~~~~~~~~
./include/linux/kernel.h:63:46: note: in definition of macro '__round_mask'
   63 | #define __round_mask(x, y) ((__typeof__(x))((y)-1))
      |                                              ^
drivers/net/ethernet/lantiq_xrx200.c:111:16: note: in expansion of macro 'round_up'
  111 |         return round_up(xrx200_max_frame_len(mtu), 4 * XRX200_DMA_BURST_LEN);
      |                ^~~~~~~~
drivers/net/ethernet/lantiq_xrx200.c:111:56: note: each undeclared identifier is reported only once for each function it appears in
  111 |         return round_up(xrx200_max_frame_len(mtu), 4 * XRX200_DMA_BURST_LEN);
      |                                                        ^~~~~~~~~~~~~~~~~~~~
./include/linux/kernel.h:63:46: note: in definition of macro '__round_mask'
   63 | #define __round_mask(x, y) ((__typeof__(x))((y)-1))
      |                                              ^
drivers/net/ethernet/lantiq_xrx200.c:111:16: note: in expansion of macro 'round_up'
  111 |         return round_up(xrx200_max_frame_len(mtu), 4 * XRX200_DMA_BURST_LEN);
      |                ^~~~~~~~
drivers/net/ethernet/lantiq_xrx200.c:112:1: error: control reaches end of non-void function [-Werror=return-type]
  112 | }
      | ^
cc1: some warnings being treated as errors
make[8]: *** [scripts/Makefile.build:280: drivers/net/ethernet/lantiq_xrx200.o] Error 1
make[7]: *** [scripts/Makefile.build:497: drivers/net/ethernet] Error 2
make[6]: *** [scripts/Makefile.build:497: drivers/net] Error 2
make[5]: *** [Makefile:1822: drivers] Error 2
make[5]: Leaving directory '/home/joerg/Downloads/openwrt6431/build_dir/target-mips_24kc_musl/linux-lantiq_xrx200/linux-5.10.87'
make[4]: *** [Makefile:22: /home/joerg/Downloads/openwrt6431/build_dir/target-mips_24kc_musl/linux-lantiq_xrx200/linux-5.10.87/.modules] Error 2
make[4]: Leaving directory '/home/joerg/Downloads/openwrt6431/target/linux/lantiq'
make[3]: *** [Makefile:11: compile] Error 2
make[3]: Leaving directory '/home/joerg/Downloads/openwrt6431/target/linux'
time: target/linux/compile#8.22#3.14#11.51
    ERROR: target/linux failed to build.
make[2]: *** [target/Makefile:30: target/linux/compile] Error 1
make[2]: Leaving directory '/home/joerg/Downloads/openwrt6431'
make[1]: *** [target/Makefile:23: /home/joerg/Downloads/openwrt6431/staging_dir/target-mips_24kc_musl/stamp/.target_compile] Error 2
make[1]: Leaving directory '/home/joerg/Downloads/openwrt6431'
make: *** [/home/joerg/Downloads/openwrt6431/include/toplevel.mk:230: world] Fehler 2
```

jospezial commented:
I had to create following patches:
target/linux/lantiq/patches-5.10/0711-net-lantiq-xrx200_Hardcode_the_burst_length_value_7e553c44f09a8f536090904c6db5b8c9dbafa03b.patch
target/linux/lantiq/patches-5.10/0712-net-lantiq-xrx200_increase_buffer_reservation_1488fc204568f707fe2a42a913788c00a95af30e.patch

in 0711-net-lantiq-xrx200_Hardcode_the_burst_length_value_7e553c44f09a8f536090904c6db5b8c9dbafa03b.patch
I had to add
```
priv->
```
to the old-line to get the patch applied.

Compile tested.

abajk commented:
> @abajk When will we see your latest commits in stable (LTS) kernel?

My patch should land in 5.17 unless something unexpected happens. [The patch](https://git.kernel.org/pub/scm/linux/kernel/git/netdev/net-next.git/commit/?id=c3e6b2c35b34214c58c1e90d65dab5f5393608e7) has been accepted by the netdev mailing list.

jekkos commented:
> baby jumbo frames run on 7412 with rfc4638 setup from July 21th . Tested VLAN bridge and non-VLAN bridge .

Could you share how you configured this ? The pppoe-wan set to 1500 and dsl.0 to 1508? or both 1508?

hauke commented:
Some of @abajk patches were accepted upstream, but they will probably not get backported to kernel 5.10, because they add new features and not only fix bugs. Does someone want to backport them to OpenWrt master?

abajk commented:
> There's a very slight issue with using the CPU port mtu to set FLEN: when setting a slave MTU, if the master device already had a large enough MTU then port_change_mtu is not called on the CPU port. This feels like it might not be intended; I'll ask on netdev.

@tomjnixon Are you planning to send your patch upstream?

abajk commented:
I think this PR can be closed. The required patches have been accepted [upstream](https://github.com/torvalds/linux/commit/c40bb4fedcd6b8b6a714da5dd466eb88ed2652d1).  [These patches](https://github.com/openwrt/openwrt/tree/master/target/linux/lantiq/patches-5.10) (patch numbers 702, 703, 704 and 706) are also available in OpenWRT.

hauke commented:
This was applied some time ago in 255268ce1a218a670e653dc5c83067f704164b7c and improved in b4970dab6b0c3e13715f4b13de42d72a74c1c9e9
